### DISCUSSION TITLE
Attributes, properties, events, methods

### DISCUSSION THREAD

**Discussion by tomcrane** [digirati-co-uk/iiif-canvas-panel#45] [General]
## Shared by `<img-service/>` and `<canvas-panel />`

| Attribute         | Property         | Description                                                        |
|-------------------|------------------|--------------------------------------------------------------------|
| render            | render           | Sets the render mode (zoom, static, responsive)                    |
| preferred-formats | preferredFormats | Sets which image format should be used when multiple are available |

## Only for `<img-service />`

| Attribute         | Property         | Description                                                        |
|-------------------|------------------|--------------------------------------------------------------------|
| src               | imageServiceUrl  | Set the source to be loaded and used                               |
| target               | target                  | Which region should be the initial starting point for viewing.  If this is changed dynamically, the viewer will transition to the new target.       |
| -                    | targetOptions           | Options for what to do with a target outside of the default behaviour. As the target is like a single annotation, this is similar to those options. |

## Only for `<canvas-panel />`

| Attribute            | Property                | Description                                                                                                                                         |
|----------------------|-------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------|
| iiif-content         | -                       | Catch all for various formats of IIIF content (URL, content state)                                                                                  |
| part-of         | -                       | Dereferencable object, used alongside iiif-content                                                                                  |
| manifest-id          | manifestId              | Changes which manifest will be loaded and used                                                                                                      |
| canvas-id            | canvasId                | Changes which canvas will be loaded and used                                                                                                        |
| -                    | annotationOptions       | Options for what to do with annotation lists found within the manifest. These could be display options like box styles.                             |
| external-annotations | loadExternalAnnotations | Boolean to tell the component if it should find and load external annotation lists and display them.                                                |
| annotation-list      | annotationList          | Link to an external annotation list to be used instead of any embedded one. This could be used similar to choice-id to switch between lists.        |
| choice-id            | choiceIds               | List of ids (comma separated in the attribute) for choices that should be pre-made. [requires knowledge of the IIIF]                                |

**Comment by stephenwf** [comment]
Diagrams (TBC)

### Fetching manifest

![image](https://user-images.githubusercontent.com/8266711/129024215-0665d061-7167-4697-b815-80f5d9fa2971.png)

I guess the HTML `attribute` could also be `part-of="-- manifest id -- "`

### Fetching canvas

[tbc]


### Rendering canvas

[tbc]

### Loading content state

[tbc]

**Comment by tomcrane** [comment]
Temporary TODO
https://digirati.slack.com/archives/D0E15T142/p1628693033052500

Reflect any decisions from that into table at top.

**Comment by stephenwf** [comment]
## Events

Events will be key to opening the canvas panel black box and enabling developers
to write supplemental code quickly to understand the content being viewed.

This is a breakdown of the existing native events from both the viewer and the
data, and works up to possible events and the formats they could take.

### Native events

When a user clicks on an image in the viewer, it will dispatch a mouse event.
Atlas works by catching all of these events on an HTML element that is the same
size as the viewport. This could be a `<div />` wrapping the viewer, or maybe a
`<canvas />` element.

Atlas takes the users mouse position, the position of that container and the
current viewport to calculate _which_ elements inside the viewer were clicked.
This is modelled after browser events, so the event makes its way _down_ into
the viewer (world, world item, image, tile, etc.) and then bubbles back up. This
event can be listened to at any level inside of Atlas and stopped. For example,
an `onClick` event on a Canvas (i.e. `WorldObject`) might stop the event being
dispatched to the images comprising it, treating it as a single unit.

When Atlas calculates the x, y position of the browser event, it adds that to
the native mouse event:

```js
e.atlas = {
  x: 123,
  y: 456,
};
```

This is the only addition to the native events currently. When working on more
abstractions, we need to decide if we add more properties to the native event,
or create new objects containing only the data required.

Another factor is the performance, specifically with events that are dispatched
more often (like mousemove). When Atlas adds it's position data it doesn't set
it like the example above, instead it shares an object:

```js
const pooledEvent = {
  x: 0,
  y: 0,
};

function someHandler(e) {
  pooledEvent.x = calculateX(e);
  pooledEvent.y = calculateY(e);
  e.atlas = pooledEvent;

  // ...
}
```

These optimisations are critical for Atlas when it's used as a lower level
building block - creating drag and drop functionality, resizing, animating etc.
With the canvas panel component we might be able to introduce constraints that
allow heavier events to occur less often.

### Data events

A data event is dispatched when a component has received some IIIF data and is
ready to start using it to render something. This is a very generic and
open-ended description to avoid semantics around loading data or to be
prescriptive on what could dispatch these events.

```js
async function renderMyComponent() {
  const range = vault.fromRef({ id: "...", type: "Range" });

  dispatchDataEvent(range); // API likely won't work exactly like this.

  // ... render thing.
}
```

Given this example, what events should `dispatchDataEvent` dispatch? To figure
that out, what might an external user want to know about the range.

- **Loaded range** - Access to the full item might be useful for syncing an
  external UI
- **Range with sub-ranges** - also useful for UIs
- **Range target** - perhaps knowing the target of a range (canvas, canvas
  region, temporal?)
- **Range behavior** - You may only want to know if a range with a specific
  behavior is loaded

In order to provide information like this about all of the different types of
IIIF resources we have a few options. Dispatch generic events with a way to
understand what's been loaded or we could dispatch generic specific events.

#### Generic events

One way to do generic events is to provide as little information as possible.

```js
document.addEventListener("range-load", (range) => {
  // range = { id: 'http://...range-1' };
  const fullRange = vault.fromRef(range.id);

  // user can now use helpers to read the range and respond.
  if (utils.rangeWithSubRanges(fullRange)) {
    /* ... */
  }

  // Or read the normalized entity.
  if (fullRange.behavior.contains("no-nav")) {
    /* ... */
  }
});
```

This would move a lot of the decision making to the user, with a lighter
abstraction.

Another way to do generic events would be to pass more information or a helper
into the event. The above example may look like this instead:

```js
document.addEventListener("range-load", (rangeHelper) => {
  // range = { id: 'http://...range-1' };
  const fullRange = rangeHelper.range;

  // user can now use helpers to read the range and respond.
  if (rangeHelper.hasSubRanges()) {
    /* ... */
  }

  // Or read the normalized entity.
  if (rangeHelper.hasBehavior("no-nav")) {
    /* ... */
  }
});
```

This would help users easily access the helpers directly without loading
something else, but would require those helpers to be baked in. A hybrid
approach could also work, with the most common helpers available in the event.

#### Specific events

Another option for dispatching events is to make very specific events for
various types of content. These would likely only be fired if a user was
listening for them.

```js
document.addEventListener("range-load:with-sub-ranges", (range) => {
  /* ... */
});

document.addEventListener("range-load:behavior-no-nav", (range) => {
  /* ... */
});

// ...
```

These would require a lot of documentation or a naming convention but would make
the purpose of these events clear.

Another approach would be to make a non-standard event listener format.

```js
element
  .query({
    type: "Range",
    options: {
      withSubRanges: true,
    },
    behavior: "no-nav",
  })
  .addEventListener("content-loaded", (range) => {
    // respond to your request here.
  });
```

This would be more like a query language for the Vault, and may be quicker to
develop. It could also provide more variation on events than the
`addEventListener` could.

### Atlas events

The other type of event that needs to be dispatched to the user is from
interactions on the viewer. As mentioned in the native events section, Atlas
provides events per-object. If canvas panel creates a world object that
represents a canvas, it would be responsible for relaying an event tying the
Atlas event to that particular canvas.

The question is how to dispatch these events and what's the best developer
experience. What use-cases can we support with them and how can we save
developers time.

Starting from Atlas itself, here is we would internally listen to a click event:

```js
worldItem.addEventListener("click", (e) => {
  e.atlas.x; // x position in world
  e.atlas.y; // y position in world
  // e is normal pointer event.
});
```

If canvas panel created this `worldItem` and wanted to dispatch it's own event,
it could look something like this.

```js
class CanvasPanel extends HTMLElement {
  createCanvas() {
    const worldItem = new WorldItem();
    // ... build up the item.

    for (const annotation of annotations) {
      const displayAnnotation = new Box();
      // .. create annotation

      displayAnnotation.addEventListener('click', () => {
        // Dispatch an Annotation event to the CanvasPanel web component
        this.dispatch('annotation-click', { annotation: annotation, canvas: this.canvas, helper: ... });
      })

      worldItem.appendChild(displayAnnotation)
    }

    worldItem.addEventListener("click", (e) => {
      // Dispatch a Canvas event to the CanvasPanel web component
      this.dispatch('canvas-click', { canvas: this.canvas, helper: ... });
    });
  }
}
```

and then the end user would be able to add event listeners from a canvas panel
object directly:

```js
cp.addEventListener("annotation-click", (e) => {
  e.annotation; // the annotation
  e.canvas; // the canvas;
  e.helper; // some helper passed in.
});
```

The same questions apply to the data. How specific should the events be. Another
option may be to use a query system.

```js
cp.query({
  type: "Annotation",
  motivation: "linking",
}).addEventListener("click", (e) => {
  // Handle the event here.
});
```
